### ISSUE TITLE
conditionally rendering interactive elements on v2.1

### ISSUE THREAD

**Issue by bcmdarroch** [chef/automate#2363]
## User Story

We use introspection in the UI to determine if certain buttons or interactive elements should be rendered, given the logged in user's permissions. 

Currently, introspection does not consider projects when returning a user's permissions. This means, if a user has permission to delete a resource in only one project, the user will be able to see all delete buttons for all the resources. The user can try to delete another resource in a project they don't have access to, but they will see an Unauthorized error.

## Definition of Done

We need to be able to hide interactive elements for resources from users based on the resource's projects and the user's permissions.

### Current Implementation

On init, the UI sends an introspect request that returns a map of all non-parameterized endpoints (i.e. not ending in a specific resource ID). This map is cached under `userperms` in the ngrx/store.

Introspection begins with a list of paths. They can be passed in from the UI (`'{"path":"/iam/v2beta/teams", "parameters":[]}'`) or we can pass all non-parameterized API paths. 
Note: The `parameters` field does not appear to work.

In the gateway, the path(s) are mapped to Action and Resource pairs. The action-resource pairs, along with the Subjects of the request, are passed along to authz-service.

In authz, OPA takes the action-resource Pairs, and Subjects and returns all allowed pairs of actions and resources (`[ iam:teams, iam:teams:create ], [ iam:teams, iam:teams:list ]`, etc). This set of action-resources pairs is then translated into a user permissions map like the one below:

```
{
  "endpoints": {
    "/iam/v2beta/teams": {
      "get": true,
      "put": false,
      "post": true,
      "delete": false,
      "patch": false
    },
    "/iam/v2beta/policies": {
      "get": false,
      "put": false,
      "post": false,
      "delete": true,
      "patch": false
    },
   ... and so on for the rest of the non-parameterized APIs
  }
}
```
Parameterized introspection works but is not widely used in the UI due to performance issues (an individual API request must be made per resource). 

### Technical Questions

1. Should we make use of the current implementation of introspection? Can we refactor it to suit our needs?
2. What should the UI be responsible for? 
3. What should authz-service be responsible for?

**Comment by bcmdarroch** [comment]
We came up with two spike docs.
- [overview of solutions](https://docs.google.com/document/d/1L-Jcn0hvNAY2qlfACYCjBp05mtU_EkmUQcO9A0OA9Po/edit?usp=sharing)
- [implementation notes](https://docs.google.com/document/d/1AfGtA593b9ouC0uuBZYXoe2wgRn1LrEWqBMb7QmbrIA/edit?usp=sharing)

We had a team discussion ([recording](https://chef.zoom.us/recording/play/K4d9-3zJvy0Ty_Y4u8CgvkmT8hwrUJmtZ3eX6S_w0_Lz-3Z8BLOfUy5NQRnKeMBH?continueMode=true)) and came up with the following summary of the proposed solutions:

**Problem Summary:** 
We want to conditionally show interactive elements given the logged-in user's project permissions, and reduce list page load times by making as few calls as possible.

**Technical Challenges:** 
OPA, which determines the introspection response, does not currently account for project permissions. It also does not have any data on what resources belong to which projects.

**Solutions boil down to:**
1. Make the UI responsible for knowing project-resource mappings. Currently, the UI has all this information, since our list APIs include project information. The work would involve updating the existing introspection endpoint and OPA to take in projects. 
2. Make authz-service responsible for knowing project-resource mappings. This would require a substantial amount of work, as the domain services would probably need to start updating this map on every resource creation/update. 

**Team TLDR;**
- The first option seems less time-consuming. But it makes less sense if we intend to stop returning project info with our List APIs.
- The second option seems more correct responsibility-wise and gives us the flexibility to change the List APIs in the future.

Note: We did not discuss whether either of these solutions would work for ingested resources.
